mempool/v1/priority_queue.go

```diff
@@ -0,0 +1,138 @@
+package v1
+
+import (
+	"container/heap"
+	"sort"
+
+	tmsync "github.com/tendermint/tendermint/internal/libs/sync"
+)
+
+var _ heap.Interface = (*TxPriorityQueue)(nil)
+
+// TxPriorityQueue defines a thread-safe priority queue for valid transactions.
+type TxPriorityQueue struct {
+	mtx tmsync.RWMutex
+	txs []*WrappedTx
+}
+
+func NewTxPriorityQueue() *TxPriorityQueue {
+	pq := &TxPriorityQueue{
+		txs: make([]*WrappedTx, 0),
+	}
+
+	heap.Init(pq)
+
+	return pq
+}
+
+// GetEvictableTx attempts to find and return a *WrappedTx than can be evicted
+// to make room for another *WrappedTx with higher priority. If no such *WrappedTx
+// is found, nil will be returned.
+func (pq *TxPriorityQueue) GetEvictableTx(priority int64) *WrappedTx {
+	pq.mtx.RLock()
+	defer pq.mtx.RUnlock()
+
+	txs := make([]*WrappedTx, len(pq.txs))
+	copy(txs, pq.txs)
+
+	sort.Slice(txs, func(i, j int) bool {
+		return txs[i].priority < txs[j].priority
+	})
```

Comment:
if the pq list is a heap and therefore sorted by priority why do we resort here?

Reply:
The heap is max-priority. As heaps are _not_ sorted and we want the _lowest_ priority element, we have to sort here.

Reply:
Not sure to call it `PriorityQueue` is proper. the element is not added/removed items like a queue (yes, it's a heap) but also the priority things. ( Only mainly use in this function but I guess it's by design.)

Reply:
Also not sure of the performance impact if the mempool is keeping full (Will that happens by design?)

Reply:
If the mempool maintains a lot of pressure, i.e. constantly full, then yes, this could have some impact. There is nothing preventing that. But for that to happen, the rate of transaction broadcasting -> checktx -> insertion has to be more than the rate of tx reaping into blocks. The operator could choose to increase mempool limits in this case.